meson.build

```diff
@@ -0,0 +1,380 @@
+project(
+  'GALAHAD',
+  'fortran',
+  version: '5.0.0',
+  license: 'BSD-3',
+  meson_version: '>= 0.61.0',
+  default_options: [
+    'c_std=c99',
+    'cpp_std=c++14',
+  ],
+)
+
+fc = meson.get_compiler('fortran')
+fs = import('fs')  # filesystem module
+
+# Each subdir() command includes a meson.build in a subdirectory
+# that adds to the list of source files.
+
+libgalahad_link_with = []
+
+# libgalahad_blas, libgalahad_lapack
+
+libgalahad_src = []
+libgalahad_blas_src = []
+libgalahad_lapack_src = []
+subdir('src/lapack')  # libgalahad_blas/lapack only contain the f90 interfaces
+libgalahad_deps = []
+blas_dep = []
+lapack_dep = []
+
+extra_args = []
+if fc.get_id() == 'nagfor'
+  extra_args += '-fixed'
+  extra_args += '-dcfuns'
+endif
+
+libblas_opt = get_option('libblas')
+if libblas_opt == [] or libblas_opt.length() != 2
+  warning('building our own BLAS; consider providing an optimized BLAS library')
+  libgalahad_blas = library('galahad_blas',
+                            sources : libgalahad_blas_src,
+                            fortran_args : extra_args,
+                            install : true)
+  libgalahad_link_with += libgalahad_blas
+else
+  libblas = fc.find_library(libblas_opt[1], dirs : libblas_opt[0], required : true)
+  blas_dep += libblas
+endif
+
+liblapack_opt = get_option('liblapack')
+if liblapack_opt == [] or liblapack_opt.length() != 2
+  warning('building our own LAPACK; consider providing an optimized LAPACK library')
+  libgalahad_lapack = library('galahad_lapack',
+                              sources : libgalahad_lapack_src,
+                              link_with : libgalahad_blas,
+                              dependencies : blas_dep,
+                              fortran_args : extra_args,
+                              install : true)
+  libgalahad_link_with += libgalahad_lapack
+else
+  liblapack = fc.find_library(liblapack_opt[1], dirs : liblapack_opt[0], required : true)
+  lapack_dep += liblapack
+endif
+
+libgalahad_deps += blas_dep + lapack_dep
+
+# Decide what goes into libgalahad_hsl, if anything, based on information
+# supplied by the user
+
+libgalahad_hsl_src = []
+
+has_hslarchive = false
+hslarchive_galahad = get_option('hslarchive-galahad')
+if hslarchive_galahad != ''
+  if fs.is_dir(hslarchive_galahad)
+    ad02_path = hslarchive_galahad / 'ad02'
+    ma27_path = hslarchive_galahad / 'ma27'
+    ma33_path = hslarchive_galahad / 'ma33'
+    has_hslarchive = true
+  else
+    error('directory not found: ', hslarchive_galahad)
+  endif
+else
+  warning('consider downloading hslarchive-galahad to improve functionality')
```

Comment:
You may want to consider using https://mesonbuild.com/Wrap-dependency-system-manual.html#specific-to-wrapfile to make this easier.

It does look like hslarchive-galahad requires an agreement to download, so you cannot take advantage of automatic wrap dependency downloads, but you can give people a location to download it to (`subprojects/packagefiles/`) and also abstract out the logic for building it into an overlay meson.build